partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rel/RelJoinOuterFull.kt

```diff
@@ -1,61 +1,105 @@
- * - If the condition passes, return the merged record (equivalent to result of INNER JOIN)
- * - If the condition does not pass, we need a way to return two records (one where the LHS is padded with nulls, and
- *   one where the RHS is padded with nulls). How we do this:
- *   - We maintain the [previousLhs] and [previousRhs]. If they are null, we then compute the next LHS and RHS. We
- *     store their values in-memory. Then we return a merged Record where the LHS is padded and the RHS is not (equivalent
- *     to result of RIGHT OUTER JOIN).
- *   - If they aren't null, then we pad the RHS with NULLS (we assume we've already padded the LHS) and return (equivalent
- *     to result of LEFT OUTER JOIN). We also make sure [previousLhs] and [previousRhs] are now null.
- *
- * Performance Analysis: Assume that [lhs] has size M and [rhs] has size N.
- * - Time: O(M * N)
- * - Space: O(1)
- */
 internal class RelJoinOuterFull(
-    override val lhs: Operator.Relation,
-    override val rhs: Operator.Relation,
-    override val condition: Operator.Expr,
-) : RelJoinNestedLoop() {
-
-    private var previousLhs: Record? = null
-    private var previousRhs: Record? = null
-
-    override fun next(): Record {
-        if (previousLhs != null && previousRhs != null) {
-            previousRhs!!.padNull()
-            val newRecord = previousLhs!! + previousRhs!!
-            previousLhs = null
-            previousRhs = null
-            return newRecord
+    private val lhs: Operator.Relation,
+    private val rhs: Operator.Relation,
+    private val condition: Operator.Expr,
+    lhsType: Rel.Type,
+    rhsType: Rel.Type
+) : RelPeeking() {
+
+    @OptIn(PartiQLValueExperimental::class)
+    private val lhsPadded = Record(
+        Array(rhsType.schema.size) { lhsType.schema[it].type.toRuntimeType().toNull().invoke() }
+    )
+
+    @OptIn(PartiQLValueExperimental::class)
+    private val rhsPadded = Record(
+        Array(rhsType.schema.size) { rhsType.schema[it].type.toRuntimeType().toNull().invoke() }
+    )
+
+    private lateinit var env: Environment
+    private lateinit var iterator: Iterator<Record>
+
+    override fun openPeeking(env: Environment) {
+        this.env = env
+        lhs.open(env)
+        iterator = implementation()
+    }
+
+    override fun peek(): Record? {
+        return when (iterator.hasNext()) {
+            true -> iterator.next()
+            false -> null
         }
-        return super.next()
+    }
+
+    override fun closePeeking() {
+        lhs.close()
+        rhs.close()
+        iterator = emptyList<Record>().iterator()
     }
 
     /**
-     * Specifically, for FULL OUTER JOIN, when the JOIN Condition ([condition]) is TRUE, we need to return the
-     * rows merged (without modification). When the JOIN Condition ([condition]) is FALSE, we need to return
-     * the LHS padded (and merged with RHS not padded) and the RHS padded (merged with the LHS not padded).
+     * FULL OUTER JOIN (CANNOT BE LATERAL)
```

Comment:
Worth to call out spec section 5.5 here: 

The FROM clause expression:
`l FULL JOIN r ON *$c$*`

replicates SQL’s FULL JOIN functionality. It assumes that (alike SQL) the lhs of r does not use variables defined from l.